### PR TITLE
End2End test with all steps

### PR DESCRIPTION
**Pull request by satween**
1. Compilation
2. Lightsaber processing
3. Verification of the errors

### CODE REVIEW

File: samples/sample-kotlin/build.gradle [hunk 1]
```diff
@@ -6,7 +6,7 @@ apply plugin: 'com.joom.lightsaber'
 sourceCompatibility = javaVersion
 targetCompatibility = javaVersion
 
-mainClassName = 'com.joom.lightsaber.sample.LightsaberSample'
+mainClassName = 'com.joom.lightsaber.testcase.provided_by_without_inject.AppMain'
```
Comment (MichaelRocks) [inline samples/sample-kotlin/build.gradle]: What's the reason for this change?

File: processor/src/main/java/com/joom/lightsaber/processor/io/JrtFileSource.kt [hunk 2]
```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright 2020 SIA Joom
```
Comment (MichaelRocks) [inline processor/src/main/java/com/joom/lightsaber/processor/io/JrtFileSource.kt]: 2022 :)

File: processor/src/test/java/com/joom/lightsaber/processor/end2end/End2EndTest.kt [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+import java.nio.file.Paths
+import kotlin.io.path.absolutePathString
+import kotlin.io.path.exists
+import kotlin.io.path.extension
+import kotlin.io.path.isDirectory
+import kotlin.streams.toList
+import org.jetbrains.kotlin.cli.common.ExitCode
+import org.jetbrains.kotlin.cli.jvm.K2JVMCompiler
+
+open class End2EndTest(private val root: Path) {
```
Comment (MichaelRocks) [inline processor/src/test/java/com/joom/lightsaber/processor/end2end/End2EndTest.kt]: I believe this class should be a `TestRule` and should be used in tests via `@Rule`. And I suggest calling it something like `IntegrationTestRule`.

File: processor/src/test/java/com/joom/lightsaber/processor/end2end/End2EndTest.kt [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+    val compiler = K2JVMCompiler()
+    // https://kotlinlang.org/docs/compiler-reference.html#common-options
+
+    val exitCode = compiler.exec(
+      System.err,
+      input,
+      "-d", compiledFilesDirectory.absolutePathString(),
+      "-cp", classpath,
+      "-jvm-target", "1.8",
```
Comment (MichaelRocks) [inline processor/src/test/java/com/joom/lightsaber/processor/end2end/End2EndTest.kt]: Shouldn't it be 9 or later with the new version of Grip?

File: processor/src/test/java/com/joom/lightsaber/processor/end2end/End2EndTest.kt [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+      "-language-version", "1.4",
```
Comment (MichaelRocks) [inline processor/src/test/java/com/joom/lightsaber/processor/end2end/End2EndTest.kt]: Just curious, why are you using such an old version?

**Comment by MichaelRocks** [comment]
Maybe it would be better to put all the test classes and interfaces into a single file to make tests easier to read?

File: processor/src/test/java/com/joom/lightsaber/processor/validation/ValidatorTest.kt [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    )
+
+    verify(reporterMock).reportError(
+      "Invalid configuration for dependency: " +
+          "test_case_projects.provided_by_without_inject.DependencyInterface. \n" +
+          "\ttest_case_projects.provided_by_without_inject.DependencyImplementation: " +
+          "Using @ProvidedBy without @Inject. You must either add at least one constructor " +
+          "with @Inject annotation or add @Provide method\n"
+    )
+  }
```
Comment (MichaelRocks) [inline processor/src/test/java/com/joom/lightsaber/processor/validation/ValidatorTest.kt]: I think that common usages should be built into the runner or provided as extensions so you can write test cases using a single command. For instance it could be:
```
assertValidProject("valid_configuration", "<optional message>")
``` 

or

```
assertInvalidProject("provided_by_without_inject", "<optional message>")
```

File: samples/sample-kotlin/build.gradle (hunk 1 above)
Comment (satween) [reply samples/sample-kotlin/build.gradle]: rolled back

File: processor/src/main/java/com/joom/lightsaber/processor/io/JrtFileSource.kt (hunk 2 above)
Comment (satween) [reply processor/src/main/java/com/joom/lightsaber/processor/io/JrtFileSource.kt]: fixed

File: processor/src/test/java/com/joom/lightsaber/processor/end2end/End2EndTest.kt (hunk 3 above)
Comment (satween) [reply processor/src/test/java/com/joom/lightsaber/processor/end2end/End2EndTest.kt]: done

File: processor/src/test/java/com/joom/lightsaber/processor/end2end/End2EndTest.kt (hunk 4 above)
Comment (satween) [reply processor/src/test/java/com/joom/lightsaber/processor/end2end/End2EndTest.kt]: it's experimental params, removed

File: processor/src/test/java/com/joom/lightsaber/processor/end2end/End2EndTest.kt (hunk 5 above)
Comment (satween) [reply processor/src/test/java/com/joom/lightsaber/processor/end2end/End2EndTest.kt]: it's experimental params, removed

File: processor/src/test/java/com/joom/lightsaber/processor/validation/ValidatorTest.kt (hunk 6 above)
Comment (satween) [reply processor/src/test/java/com/joom/lightsaber/processor/validation/ValidatorTest.kt]: done

File: processor/src/main/java/com/joom/lightsaber/processor/analysis/Analyzer.kt [hunk 7]
```diff
@@ -20,29 +20,35 @@ import com.joom.grip.Grip
...
 import java.io.File
+import java.nio.file.Path
 
 class Analyzer(
   private val grip: Grip,
   private val errorReporter: ErrorReporter,
   private val projectName: String
 ) {
 
+  @Deprecated("Deprecated.", ReplaceWith("analyzePaths(files.map{ it.toPath() })"))
```
Comment (MichaelRocks) [inline processor/src/main/java/com/joom/lightsaber/processor/analysis/Analyzer.kt]: I don't think we really need to preserve the legacy one. Let's just change its signature.

Comment (satween) [reply processor/src/main/java/com/joom/lightsaber/processor/analysis/Analyzer.kt]: rolled back

File: processor/src/main/java/com/joom/lightsaber/processor/LightsaberParameters.kt [hunk 8]
```diff
@@ -17,12 +17,33 @@
...
+  val inputs: List<Path>,
+  val outputs: List<Path>,
+  val classpath: List<Path>,
+  val bootClasspath: List<Path>,
+  val gen: Path,
+  val projectName: String,
+  val errorReporter: ErrorReporter = ErrorReporterImpl()
+) {
+
+  constructor(
```
Comment (MichaelRocks) [inline processor/src/main/java/com/joom/lightsaber/processor/LightsaberParameters.kt]: I think we can drop this one. Anyway its ABI isn't backward compatible.

File: processor/src/test/java/com/joom/lightsaber/processor/integration/IntegrationTestRule.kt [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+      it.extension.endsWith("jar", ignoreCase = true)
+    }
+  }
+
+  private fun getFilesThat(path: Path, filter: (Path) -> Boolean): List<Path> {
+    Files.walk(path).use { paths ->
+      return paths
+        .filter(filter)
+        .toList()
+
```
Comment (MichaelRocks) [inline processor/src/test/java/com/joom/lightsaber/processor/integration/IntegrationTestRule.kt]: Redundant empty line.

**Review by MichaelRocks — APPROVED** [review]

File: processor/src/test/java/com/joom/lightsaber/processor/integration/IntegrationTestRule.kt (hunk 9 above)
Comment (satween) [reply processor/src/test/java/com/joom/lightsaber/processor/integration/IntegrationTestRule.kt]: fixed

File: processor/src/main/java/com/joom/lightsaber/processor/LightsaberParameters.kt (hunk 8 above)
Comment (satween) [reply processor/src/main/java/com/joom/lightsaber/processor/LightsaberParameters.kt]: done
